### PR TITLE
feat(google): support requester pays by exposing a function to give b…

### PR DESCRIPTION
**Pull request by Avantol13**
### New Features
- New function for assigning a SA billing access to a Google Project 
- New option to provide a userProject to signed urls to support requester pays

### CODE REVIEW

File: cirrus/google_cloud/manager.py [hunk 1]
```diff
@@ -917,6 +934,94 @@ def delete_service_account_key(self, account, key_name):
...
+        self, account_id, project_id=None, billing_role_name=None
+    ):
+        """
+        Give the specified service account id permissions to bill to the provided project
+        id. Project id will default to the current project.
+
+        Args:
+            account_id (str): email address or the uniqueId of the service account
+            project_id (str): Google project identifier
+        """
```
Comment (paulineribeyre) [inline cirrus/google_cloud/manager.py]: you're missing `billing_role_name` in the docstring
and what is the use case for providing a custom name?

File: cirrus/google_cloud/manager.py [hunk 2]
```diff
@@ -917,6 +934,94 @@ def delete_service_account_key(self, account, key_name):
...
+        full_billing_role_resource = "projects/{}/roles/{}".format(
+            project_id, billing_role_id
+        )
+
+        # get project IAM policy and see if the sa already has the necessary access
+        policy = self._get_project_iam_policy(project_id)
+        role = GooglePolicyRole(name=full_billing_role_resource)
+        roles = {role.name: role for role in policy.roles}
+        if full_billing_role_resource in roles.keys():
+            role = roles[full_billing_role_resource]
```
Comment (paulineribeyre) [inline cirrus/google_cloud/manager.py]: you could simplify this with
```
for role in policy.roles:
  if role.name == full_billing_role_resource:
```

File: cirrus/google_cloud/manager.py [hunk 3]
```diff
@@ -917,6 +934,94 @@ def delete_service_account_key(self, account, key_name):
...
+
+        # create new role with just billing access if doesn't exist already
+        # https://cloud.google.com/iam/docs/creating-custom-roles#iam-custom-roles-create-rest
+        new_role = {
+            "role_id": billing_role_id,
+            "role": {
+                "name": "",
+                "title": "Project Billing User",
+                "description": "This role grants a user access to use billing for the given resource.",
+                "included_permissions": "serviceusage.services.use",
```
Comment (paulineribeyre) [inline cirrus/google_cloud/manager.py]: how did you choose `serviceusage.services.use`? i don't see in the doc that it gives permission to bill the project and nothing else

File: cirrus/google_cloud/iam.py [hunk 4]
```diff
@@ -63,6 +68,23 @@ def __str__(self):
...
+    def get_dict(self):
+        """
+        Return representation of object
+
+        Returns:
+            str: Representation of the Policy which can be POSTed to Google's API
+        """
+        output_dict = dict()
+        output_dict["policy"] = dict()
+        output_dict["policy"]["bindings"] = [
```
Comment (paulineribeyre) [inline cirrus/google_cloud/iam.py]: you could just do
```
return {
  "policy": {
    "bindings": ...
    "etag": ...
    "version": ...
  }
}
```

File: cirrus/google_cloud/iam.py [hunk 5]
```diff
@@ -213,7 +228,13 @@ def __str__(self):
         Returns:
             str: Representation of the Role for Google's API
         """
-        return "{}{}".format(GooglePolicyRole.ROLE_PREFIX, self.name)
+        # / means the role already has a prefix in the name
+        if "/" in self.name:
```
Comment (paulineribeyre) [inline cirrus/google_cloud/iam.py]: why did you move this check from `__init__ ` to `__str__ `?
you also removed the `.strip()`s in `__init__` so maybe the name can be set incorrectly?

I think this test should also check that `name` is set correctly: https://github.com/uc-cdis/cirrus/blob/master/test/test_google_cloud_iam.py#L9

File: cirrus/google_cloud/iam.py [hunk 6]
```diff
@@ -196,13 +218,6 @@ def __init__(self, name):
         Args:
             name (str): The name of the Google role
         """
-        # If the name provided already starts with the prefix, remove it
-        if (
-            name.strip()[: len(GooglePolicyRole.ROLE_PREFIX)]
-            == GooglePolicyRole.ROLE_PREFIX
-        ):
-            name = name.strip()[len(GooglePolicyRole.ROLE_PREFIX) :]
```
Comment (paulineribeyre) [inline cirrus/google_cloud/iam.py]: if you want to put this back (see my other comment) what do you think of this?
```
self.name = self.name.strip()
if self.name.startsWith(GooglePolicyRole.ROLE_PREFIX):
  self.name = self.name[len(GooglePolicyRole.ROLE_PREFIX):]
```

File: cirrus/google_cloud/iam.py (hunk 4 above)
Comment (Avantol13) [reply cirrus/google_cloud/iam.py]: yeah, for consistency I was matching the existing function

File: cirrus/google_cloud/manager.py (hunk 1 above)
Comment (Avantol13) [reply cirrus/google_cloud/manager.py]: just in case someone has a preference as to the role name in the future, ends up being an API call change in fence instead of another release in cirrus. :man_shrugging: not a big deal, could just leave it hard-coded for now

File: cirrus/google_cloud/manager.py (hunk 3 above)
Comment (Avantol13) [reply cirrus/google_cloud/manager.py]: https://cloud.google.com/storage/docs/access-control/iam-console
`All actions that include a billing project in the request require serviceusage.services.use permission for the project that's specified.`

File: cirrus/google_cloud/manager.py (hunk 2 above)
Comment (Avantol13) [reply cirrus/google_cloud/manager.py]: good point

File: cirrus/google_cloud/iam.py (hunk 4 above)
Comment (paulineribeyre) [reply cirrus/google_cloud/iam.py]: oh i see, then maybe
- simplify `GooglePolicy.get_dict`
- make `GooglePolicy.__str__()` call `GooglePolicy.get_dict`
- simplify `GooglePolicyBinding.get_dict` if you want to :p

File: cirrus/google_cloud/iam.py (hunk 5 above)
Comment (Avantol13) [reply cirrus/google_cloud/iam.py]: google custom roles require a different prefix with "/", being "projects/project-name/roles/". so the first change was to stop trying to be smart about the role prefix during initialization and just set it to what was given. then, when converting to a valid string, if it's a normal role (e.g. no "/") preprend the necessary prefix

Comment (Avantol13) [reply cirrus/google_cloud/iam.py]: that being said, if we want to maintain exact backwards compatibility, we could add the code back in that does the initial name sanitation (removing the prefix), but we'd still need this chunk of code

File: cirrus/google_cloud/iam.py (hunk 6 above)
Comment (Avantol13) [reply cirrus/google_cloud/iam.py]: I'll do something like this to maintain as much backwards compatibility as possible (and add a clarifying comment)

File: cirrus/google_cloud/iam.py [hunk 7]
```diff
@@ -218,7 +218,15 @@ def __init__(self, name):
         Args:
             name (str): The name of the Google role
         """
-        self.name = name
+        # if using a traditional role, remove the prefix for the name
+        # NOTE: Custom roles have a different prefix, and we will transparently
+        #       have that as the name since the prefix is dynamic (e.g. it changes
+        #       based on the project/org the custom role was defined in)
+        if name.strip().startswith(GooglePolicyRole.ROLE_PREFIX):
+            self.name = name[len(GooglePolicyRole.ROLE_PREFIX) :]
```
Comment (paulineribeyre) [inline cirrus/google_cloud/iam.py]: if `name` starts with whitespace this won't work, i think you can't avoid doing `self.name = self.name.strip()` first :p

File: cirrus/google_cloud/iam.py (hunk 4 above)
Comment (Avantol13) [reply cirrus/google_cloud/iam.py]: oh, :man_facepalming: yeah good point

File: cirrus/google_cloud/manager.py [hunk 8]
```diff
@@ -956,16 +958,15 @@ def give_service_account_billing_access(
         # get project IAM policy and see if the sa already has the necessary access
         policy = self._get_project_iam_policy(project_id)
         role = GooglePolicyRole(name=full_billing_role_resource)
```
Comment (paulineribeyre) [inline cirrus/google_cloud/manager.py]: i think `role` is not used anywhere, and it's getting overwritten by the `for` loop anyway

Comment (Avantol13) [reply cirrus/google_cloud/manager.py]: oh, yup, gets created further down

File: cirrus/google_cloud/iam.py (hunk 7 above)
Comment (Avantol13) [reply cirrus/google_cloud/iam.py]: oh, shoot :man_facepalming: ugh, yeah

Comment (Avantol13) [reply cirrus/google_cloud/iam.py]: well, I can just `name = name.strip()`
